src/common/geo/GeoUtils.h

```diff
@@ -0,0 +1,124 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#pragma once
+
+#include <s2/s2loop.h>
+#include <s2/s2polygon.h>
+
+#include "common/base/StatusOr.h"
+#include "common/datatypes/Geography.h"
+#include "common/geo/io/Geometry.h"
+
+namespace nebula {
+
+class GeoUtils final {
+ public:
+  static std::unique_ptr<S2Region> s2RegionFromGeomtry(const Geometry* geom) {
+    switch (geom->shape()) {
+      case GeoShape::POINT: {
+        const auto* point = static_cast<const Point*>(geom);
+        auto s2Point = s2PointFromCoordinate(point->coord);
+        return std::make_unique<S2PointRegion>(s2Point);
+      }
+      case GeoShape::LINESTRING: {
+        const auto* lineString = static_cast<const LineString*>(geom);
+        auto coordList = lineString->coordList;
+        DCHECK_GE(coordList.size(), 2);
+        removeAdjacentDuplicateCoordinates(coordList);
+        if (coordList.size() < 2) {
+          LOG(INFO)
+              << "Invalid LineString, adjacent coordinates must not be identical or antipodal.";
+          return nullptr;
+        }
+
+        auto s2Points = s2PointsFromCoordinateList(coordList);
+        auto s2Polyline = std::make_unique<S2Polyline>(s2Points, S2Debug::DISABLE);
+        if (!s2Polyline->IsValid()) {
+          return nullptr;
+        }
+        return s2Polyline;
+      }
+      case GeoShape::POLYGON: {
+        const auto* polygon = static_cast<const Polygon*>(geom);
+        uint32_t numRings = polygon->numRings();
+        std::vector<std::unique_ptr<S2Loop>> s2Loops;
+        s2Loops.reserve(numRings);
+        for (size_t i = 0; i < numRings; ++i) {
+          auto coordList = polygon->coordListList[i];
+          DCHECK_GE(coordList.size(), 4);
+          DCHECK(isLoopClosed(coordList));
```

Comment:
It should be the normal error handling process. such as:

```c++
if (UNLIKELY(isLoopClosed(coordList))) {
  LOG(ERROR) << "";
  return nullptr;
}
```

Reply:
ok